### PR TITLE
fix: Perf of NavView/ItemsRepeater

### PR DESCRIPTION
**Pull request by dr1rrb**
fixes https://github.com/unoplatform/uno/issues/5172
#5109 

## What is the current behavior?
`NavigationView` is slow on Uno.Gallery

## What is the new behavior?
Slightly more performant

## PR Checklist
- [ ] ~~Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)~~
- [ ] ~~[Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)~~
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes
- [x] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [x] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

### CODE REVIEW

File: src/Uno.UI/UI/Xaml/Controls/Grid/Grid.Layout.cs [hunk 1]
```diff
@@ -980,9 +980,17 @@ private static double GetAvailableSizeForPosition(Memory<double> calculatedPixel
 		/// </summary>
 		private static double GetAdjustmentForSpacing(int rowOrColumnSpan, double spacing) => spacing * (rowOrColumnSpan - 1);
 
-		private static Memory<ViewPosition> FindStarSizeChildren(Span<ViewPosition> positions, Memory<ViewPosition> pixelSizeChildren, List<ViewPosition> autoSizeChildren)
+		private ViewPosition[] _resCache;
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/Grid/Grid.Layout.cs:983]: move to the top

File: src/Uno.UI/UI/Xaml/Controls/Grid/Grid.Layout.cs [hunk 2]
```diff
@@ -980,9 +980,17 @@ private static double GetAvailableSizeForPosition(Memory<double> calculatedPixel
...
+
+		private Memory<ViewPosition> FindStarSizeChildren(Span<ViewPosition> positions, Memory<ViewPosition> pixelSizeChildren, List<ViewPosition> autoSizeChildren)
 		{
-			var res = new Memory<ViewPosition>(new ViewPosition[positions.Length]);
+			if ((_resCache?.Length ?? -1) != positions.Length)
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/Grid/Grid.Layout.cs:987]: I'm wondering if we should use a ArrayPool instead.

Comment (dr1rrb) [reply src/Uno.UI/UI/Xaml/Controls/Grid/Grid.Layout.cs]: I used both: I really expect the `_resCache` to never change during the lifetime of a `Grid`, so we can avoid the cost of lookup in the pool.

File: src/Uno.UI/Microsoft/UI/Xaml/Controls/Repeater/ViewportManagerWithPlatformFeatures.cs [hunk 3]
```diff
@@ -255,7 +255,9 @@ public override void OnLayoutChanged(bool isVirtualizing)
 			{
 				m_effectiveViewportChangedRevoker?.Dispose();
 			}
-			else if (m_effectiveViewportChangedRevoker == null)
+			else if (m_effectiveViewportChangedRevoker == null
+				// Uno workaround: [Perf] Do not listen for viewport update if nothing to render!
+				&& m_owner.ItemsSourceView?.Count > 0)
```
Comment (davidjohnoliver) [inline src/Uno.UI/Microsoft/UI/Xaml/Controls/Repeater/ViewportManagerWithPlatformFeatures.cs:260]: Will this be called again if/when `ItemsSourceView.Count` becomes non-zero?

File: src/Uno.UI/Microsoft/UI/Xaml/Controls/Repeater/ViewportManagerWithPlatformFeatures.cs [hunk 4]
```diff
@@ -489,6 +491,12 @@ void EnsureScroller()
 			{
 				ResetScrollers();
 
+				// Uno workaround: [Perf] Do not listen for viewport update if nothing to render!
+				if (m_owner.ItemsSourceView?.Count <= 0)
```
Comment (davidjohnoliver) [inline src/Uno.UI/Microsoft/UI/Xaml/Controls/Repeater/ViewportManagerWithPlatformFeatures.cs:495]: same question here

File: src/Uno.UI/Microsoft/UI/Xaml/Controls/Repeater/ViewportManagerWithPlatformFeatures.cs (hunk 3 above)
Comment (dr1rrb) [reply src/Uno.UI/Microsoft/UI/Xaml/Controls/Repeater/ViewportManagerWithPlatformFeatures.cs:260]: Good question, and the answer is yes! The collection changed will trigger an `InvalidateMeasure` in all path, which will call the `OnBeginMeasure` on this class, which will call the `EnsureScroller()` (i.e. you second comment)

**Review by ajpinedam — APPROVED** [review]

**Pull request merged**